### PR TITLE
ABU-674: Locking as optional utility

### PR DESCRIPTION
**Pull request by oliverfoster**
* changes Adapt object from _.extend({}, Backbone.Events) to Backbone.Model to accommodate run-time variables in a consistent way
* extended Backbone.Model to include optional attribute locking, the locking mechanism is one lock per plugin name, to set or unset
* added _canScroll to Adapt model to block Adapt.scrollTo function from performing the scroll (useful for block sliders when they should manage their own scrolling)
* made Adapt.get("_canScroll") a locked attribute
* made Adapt.router.get("_canNavigate") a locked attribute (to overcome brians issue (https://github.com/adaptlearning/adapt_framework/issues/587)
* extended router to block all navigation events when using Adapt.router.set("_canNavigate", false, { pluginName: "plugin" });
* made router block navigation events when content is loading (to solve media element issue in ie8 on a navigate+back in quick succession)
* kept .set(attributeName, value, options) to old backboneModel.js implementation using {pluginName: ''}

This is the long awaited re-implementation of backboneModel.js. Rather than being a mediator, this is a locking mechanism. It is not enabled for every model, only for models where is it explicitly enabled. It is enabled on the Adapt and Adapt.router models at present, for _canScroll and _canNavigate respectively.

New backbone model functions can be found in https://github.com/adaptlearning/adapt_framework/blob/ABU-674/src/core/js/models/lockingModel.js

Description of new Backbone.Model functions:

```
set(attrName, attrVal, options)  : an extension to the original function

  Model.set("_canNavigate", false, { pluginName: "plugin" });
  Model.set("_canNavigate", true, { pluginName: "plugin" });


setLocking(attrName, defaultLockValue) : add locking to a model attribute

    Model.setLocking("_isVisible", false);


unsetLocking(attrName) : remove locking from a model attribute

    Model.unsetLocking("_isVisible");


isLocking(attrName) : returns t/f if the attribute is a locking attribute

    Model.isLocking("_isVisible");


isLocked(attributeName, options) : returns t/f/undefined is the attribute it locked, if it is locked for the supplied pluginName or undefined if it is not a locking attribute

    Model.isLocked("_isVisible");
    Model.isLocked("_isVisible", {pluginName: "plugin"});


getLockCount(attributeName, options) : returns an integer/undefined for the number of locks held in total for the attribute, or for the supplied pluginName or undefined if the attribute is not a locking attribute

    Model.getLockValue("_isVisible");
    Model.getLockValue("_isVisible", {pluginName: "plugin"});


setLockState(attributeName, value, options) : sets the locking value for the given pluginName to value (t/f)

    Model.setLockValue("_isVisible", true, {pluginName: "plugin"});


```

### CODE REVIEW

File: src/core/js/models/lockingModel.js [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+			if (stopProcessing) return set.apply(this, arguments);
+
+			var isSettingValueForSpecificPlugin = options && options.pluginName;
+			if (!isSettingValueForSpecificPlugin) {
+				console.error("Must supply a pluginName to change a locked attribute");
+				options.pluginName = "compatibility";
+			}
+
+			var pluginName  = options.pluginName;
+			var lockingValue = this.lockedAttributes[attrName] = !this.defaults[attrName];
```
Comment (tomgreenfield) [inline src/core/js/models/lockingModel.js]: Add existence check for defaults

File: src/core/js/models/lockingModel.js [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+		isLocking: function(attrName) {
+			var isCheckingGeneralLockingState = attrName === undefined;
+			var isUsinglockedAttributes = this.lockedAttributes !== undefined;
+
+			if (isCheckingGeneralLockingState) {
+				return isUsinglockedAttributes;
+			}
+
+			if (!isUsinglockedAttributes) return false;
```
Comment (tomgreenfield) [inline src/core/js/models/lockingModel.js]: `isUsinglockedAttributes` --> `isUsingLockedAttributes`

File: src/core/js/models/lockingModel.js [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+				return this._lockedAttributesValues[attrName][options.pluginName] ? 1 : 0;
+			}
+
+			var lockingAttributeValues = _.values(this._lockedAttributesValues[attrName]);
+			var lockingAttributeValuesSum = _.reduce(lockingAttributeValues, function(memo, value){ return memo + (value ? 1 : 0); }, 0);
+			
+			return lockingAttributeValuesSum;
+		},
+
+		setLockValue: function(attrName, value, options) {
```
Comment (tomgreenfield) [inline src/core/js/models/lockingModel.js]: Consider renaming to something like setLockState?

File: src/core/js/models/lockingModel.js [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+		isLocked: function(attrName, options) {
+			if (!(options && options.skipcheck)) { 
+				var stopProcessing =  !this.isLocking(attrName);
+				if (stopProcessing) return;
+			}
+
+			return this.getLockValue(attrName) > 0;
+		},
+
+		getLockValue: function(attrName, options) {
```
Comment (tomgreenfield) [inline src/core/js/models/lockingModel.js]: Consider renaming to something like getLockCount?

File: src/core/js/models/lockingModel.js [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+			if (this._lockedAttributesValues[attrName] === undefined) {
+				this._lockedAttributesValues[attrName] = {};	
+			}
+
+			return true;
+		},
+
+		isLocked: function(attrName, options) {
+			if (!(options && options.skipcheck)) {
```
Comment (tomgreenfield) [inline src/core/js/models/lockingModel.js]: Move `options && options.skipcheck` into shouldSkip variable?

**Comment by tomgreenfield** [comment]
+1

**Comment by moloko** [comment]
+1

**Comment by taylortom** [comment]
This seems like too big a change to go into master?

**Comment by oliverfoster** [comment]
in terms of the number of untested lines? the breadth of it's footprint in adapt? how it changes adapt's behaviour?

**Comment by moloko** [comment]
seems fine to me, there are no breaking changes are there?

**Comment by oliverfoster** [comment]
noo, no breakings.

**Comment by oliverfoster** [comment]
this is in develop already.

**Pull request closed without merging**